### ISSUE TITLE
Incosistent names during elaboration

### ISSUE THREAD

**Issue by liesnikov** [liesnikov/extensible-elaborator#11]
Seemingly an issue concentrated around https://github.com/liesnikov/extensible-elaborator/blob/68282a59d0f29725f5f489cb8228cae9750b00d1/eetc/src/TypeCheck/Elaborator.hs#L267-L281

Because we generate three names, first in generating the pi type, then when unbinding lambda and then when unbinding the pi type.

```
eetc ⊢cabal build; cabal exec eetc -- ./pi/test/IdTest.pi
Build profile: -w ghc-9.2.8 -O1
In order, the following will be built (use -v for more details):
 - eetc-0.2 (lib) (file src/Syntax/PrettyPrintInternal.hs changed)
 - eetc-0.2 (exe:eetc) (dependency rebuilt)
Preprocessing library for eetc-0.2..
Building library for eetc-0.2..
[12 of 42] Compiling Syntax.PrettyPrintInternal ( src/Syntax/PrettyPrintInternal.hs, /home/bohdan/delft/extended-elab/extended-elab/eetc/dist-newstyle/build/x86_64-linux/ghc-9.2.8/eetc-0.2/build/Syntax/PrettyPrintInternal.o, /home/bohdan/delft/extended-elab/extended-elab/eetc/dist-newstyle/build/x86_64-linux/ghc-9.2.8/eetc-0.2/build/Syntax/PrettyPrintInternal.dyn_o )
Preprocessing executable 'eetc' for eetc-0.2..
Building executable 'eetc' for eetc-0.2..
Linking /home/bohdan/delft/extended-elab/extended-elab/eetc/dist-newstyle/build/x86_64-linux/ghc-9.2.8/eetc-0.2/x/eetc/build/eetc/eetc ...
processing IdTest.pi...
Parsing File "./pi/test/IdTest.pi"
elaborating...
Elaborating module "IdTest"
warning: generated a new renaming, old name id new name id
warning: ./pi/test/IdTest.pi:3:6:
  generated a new renaming, old name z3 new name z4
  In the expression
    (z : Type) -> z -> z
warning: ./pi/test/IdTest.pi:3:18:
  generated a new renaming, old name _7 new name _8
  In the expression
    z -> z
warning: ./pi/test/IdTest.pi:4:6:
  generated a new renaming, old name x20 new name x22
  In the expression
    \x y. y
warning: ./pi/test/IdTest.pi:4:6:
  generated a new renaming, old name y28 new name y30
  In the expression
    \x y. y
warning: After checking an entry there are unsolved constraints
With current active constraints being
  {  32 : constraint 32: (?_23 [ id0  →  id0, x22  →  x22 ] ~ ?_25 [ _24  →  y30, id0  →  id0, x22  →  x22 ]) : Type,
     43 : constraint 43: (_16 ~ ?_23 [ x22  →  x22, id0  →  id0 ]) : Type,
     46 : constraint 46: (_16 ~ ?_25 [ _24  →  _44, x22  →  x22, id0  →  id0 ]) : Type  }
And current solutions to metas being
  {  ?_1 : Type,
     ?_5 : Type,
     ?_9 : Type,
     ?_11 : Type,
     ?_13 : Type,
     ?_15 : Type,
     ?_17 : _16 -> _16,
     ?_18 : ?_33 [ ] -> ?_36 [ ],
     ?_26 : ?_42 [ ] -> ?_45 [ ],
     ?_33 : Type,
     ?_36 : _16 -> _16  }
Solved constraints are
  {  2 : constraint 2: (Type ~ Type) : Type,
    6 : constraint 6: (Type ~ Type) : Type,
    10 : constraint 10: (Type ~ Type) : Type,
    12 : constraint 12: (Type ~ Type) : Type,
    14 : constraint 14: (Type ~ Type) : Type,
    19 : constraint 19: ((z : Type) -> z0 -> z0 ~ (_ : ?_15 [ id0  →  id0 ]) -> ?_17 [ _16  →  _0, id0  →  id0 ]) : Type,
    27 : constraint 27: (?_17 [ _16  →  x22, id0  →  id0 ] ~ (_ : ?_23 [ x22  →  x22, id0  →  id0 ]) -> ?_25 [ _24  →  _0, x22  →  x22, id0  →  id0 ]) : Type,
    34 : constraint 34: (Type ~ ?_15 [ id0  →  id0 ]) : Type,
    37 : constraint 37: (z35 -> z35 ~ ?_17 [ _16  →  z35, id0  →  id0 ]) : Type,
    41 : constraint 41: (_16 -> _16 ~ (_ : ?_23 [ x22  →  x22, id0  →  id0 ]) -> ?_25 [ _24  →  _0, x22  →  x22, id0  →  id0 ]) : Type  }
Blocked constraints are
  {   }
warning: generated a new renaming, old name id2 new name id247
warning: ./pi/test/IdTest.pi:6:7:
  generated a new renaming, old name y50 new name y51
  In the expression
    (y : Type) -> y -> id Type y
warning: ./pi/test/IdTest.pi:6:19:
  generated a new renaming, old name _54 new name _55
  In the expression
    y -> id Type y
warning: ./pi/test/IdTest.pi:7:7:
  generated a new renaming, old name x81 new name x83
  In the expression
    \x y. y
warning: ./pi/test/IdTest.pi:7:7:
  generated a new renaming, old name y89 new name y91
  In the expression
    \x y. y
warning: After checking an entry there are unsolved constraints
With current active constraints being
  {  32 : constraint 32: (?_23 [ id0  →  id0, x22  →  x22 ] ~ ?_25 [ _24  →  y30, id0  →  id0, x22  →  x22 ]) : Type, 43 : constraint 43: (_16 ~ ?_23 [ x22  →  x22, id0  →  id0 ]) : Type, 46 : constraint 46: (_16 ~ ?_25 [ _24  →  _44, x22  →  x22, id0  →  id0 ]) : Type, 93 : constraint 93: (?_84 [ id247  →  id247, x83  →  x83 ] ~ ?_86 [ _85  →  y91, id247  →  id247, x83  →  x83 ]) : Type, 198 : constraint 198: (_77 ~ ?_84 [ x83  →  x83, id247  →  id247 ]) : Type, 201 : constraint 201: (_77 ~ ?_86 [ _85  →  _199, x83  →  x83, id247  →  id247 ]) : Type, 206 : constraint 206: (_57 ~ Type) : Type, 211 : constraint 211: (_57 ~ Type) : Type  }
And current solutions to metas being
  {  ?_1 : Type, ?_5 : Type, ?_9 : Type, ?_11 : Type, ?_13 : Type, ?_15 : Type, ?_17 : _16 -> _16, ?_18 : ?_33 [ ] -> ?_36 [ ], ?_26 : ?_42 [ ] -> ?_45 [ ], ?_33 : Type, ?_36 : _16 -> _16, ?_48 : Type, ?_52 : Type, ?_56 : Type, ?_58 : _57 -> _57, ?_59 : ?_106 [ ] -> ?_109 [ ], ?_61 : Type, ?_63 : Type, ?_65 : Type, ?_66 : ?_188 [ ] -> ?_191 [ ], ?_68 : Type, ?_70 : Type, ?_72 : Type, ?_74 : Type, ?_76 : Type, ?_78 : _77 -> _77, ?_79 : ?_140 [ ] -> ?_143 [ ], ?_87 : ?_197 [ ] -> ?_200 [ ], ?_106 : Type, ?_109 : _57 -> _57, ?_140 : Type, ?_143 : _77 -> _77  }
Solved constraints are
  {  2 : constraint 2: (Type ~ Type) : Type, 6 : constraint 6: (Type ~ Type) : Type, 10 : constraint 10: (Type ~ Type) : Type, 12 : constraint 12: (Type ~ Type) : Type, 14 : constraint 14: (Type ~ Type) : Type, 19 : constraint 19: ((z : Type) -> z0 -> z0 ~ (_ : ?_15 [ id0  →  id0 ]) -> ?_17 [ _16  →  _0, id0  →  id0 ]) : Type, 27 : constraint 27: (?_17 [ _16  →  x22, id0  →  id0 ] ~ (_ : ?_23 [ x22  →  x22, id0  →  id0 ]) -> ?_25 [ _24  →  _0, x22  →  x22, id0  →  id0 ]) : Type, 34 : constraint 34: (Type ~ ?_15 [ id0  →  id0 ]) : Type, 37 : constraint 37: (z35 -> z35 ~ ?_17 [ _16  →  z35, id0  →  id0 ]) : Type, 41 : constraint 41: (_16 -> _16 ~ (_ : ?_23 [ x22  →  x22, id0  →  id0 ]) -> ?_25 [ _24  →  _0, x22  →  x22, id0  →  id0 ]) : Type, 49 : constraint 49: (Type ~ Type) : Type, 53 : constraint 53: (Type ~ Type) : Type, 60 : constraint 60: ((z : Type) -> z0 -> z0 ~ (_ : ?_56 [ _55  →  _55, y51  →  y51 ]) -> ?_58 [ _57  →  _0, _55  →  _55, y51  →  y51 ]) : Type, 62 : constraint 62: (Type ~ ?_56 [ _55  →  _55, y51  →  y51 ]) : Type, 67 : constraint 67: (?_58 [ _57  →  Type, _55  →  _55, y51  →  y51 ] ~ (_ : ?_63 [ _55  →  _55, y51  →  y51 ]) -> ?_65 [ _64  →  _0, _55  →  _55, y51  →  y51 ]) : Type, 69 : constraint 69: (Type ~ ?_63 [ _55  →  _55, y51  →  y51 ]) : Type, 71 : constraint 71: (?_65 [ _64  →  y51, _55  →  _55, y51  →  y51 ] ~ Type) : Type, 73 : constraint 73: (Type ~ Type) : Type, 75 : constraint 75: (Type ~ Type) : Type, 80 : constraint 80: ((y : Type) -> y0 -> id0 Type y0 ~ (_ : ?_76 [ id247  →  id247 ]) -> ?_78 [ _77  →  _0, id247  →  id247 ]) : Type, 88 : constraint 88: (?_78 [ _77  →  x83, id247  →  id247 ] ~ (_ : ?_84 [ x83  →  x83, id247  →  id247 ]) -> ?_86 [ _85  →  _0, x83  →  x83, id247  →  id247 ]) : Type, 107 : constraint 107: (Type ~ ?_56 [ _55  →  _55, y51  →  y51 ]) : Type, 110 : constraint 110: (z108 -> z108 ~ ?_58 [ _57  →  z108, _55  →  _55, y51  →  y51 ]) : Type, 141 : constraint 141: (Type ~ ?_76 [ id247  →  id247 ]) : Type, 144 : constraint 144: (y142 -> id0 Type y142 ~ ?_78 [ _77  →  y142, id247  →  id247 ]) : Type, 151 : constraint 151: (Type ~ Type) : Type, 163 : constraint 163: (_57 -> _57 ~ (_ : ?_63 [ _55  →  _55, y51  →  y51 ]) -> ?_65 [ _64  →  _0, _55  →  _55, y51  →  y51 ]) : Type, 179 : constraint 179: (_77 -> _77 ~ (_ : ?_84 [ x83  →  x83, id247  →  id247 ]) -> ?_86 [ _85  →  _0, x83  →  x83, id247  →  id247 ]) : Type, 189 : constraint 189: (_57 ~ ?_63 [ _55  →  _55, y51  →  y51 ]) : Type, 192 : constraint 192: (_57 ~ ?_65 [ _64  →  _190, _55  →  _55, y51  →  y51 ]) : Type  }
Blocked constraints are
  {   }
type checking...
Checking module "IdTest"
module IdTest where
id : (z : Type) -> z0 -> z0
id = \x y. y0
id2 : (y : Type) -> y0 -> id0 Type y0
id2 = \x y. y0
```

**Comment by liesnikov** [comment]
partially fixed by e8bc9ef59a4b0ecf0b9b213195c8e7ce7db002ae

now the first declaration in idtest elaborates just fine, but the second one runs into a similar issue

```
Elaborating module "IdTest"
warning: generated a new renaming, old name id new name id
warning: ./pi/test/IdTest.pi:3:6:
  generated a new renaming, old name z3 new name z4
  In the expression
    (z : Type) -> z -> z
warning: ./pi/test/IdTest.pi:3:18:
  generated a new renaming, old name _7 new name _8
  In the expression
    z -> z
warning: ./pi/test/IdTest.pi:4:6:
  generated a new renaming, old name x15 new name x16
  In the expression
    \x y. y
warning: ./pi/test/IdTest.pi:4:6:
  generated a new renaming, old name y22 new name y23
  In the expression
    \x y. y
warning: generated a new renaming, old name id2 new name id246
warning: ./pi/test/IdTest.pi:6:7:
  generated a new renaming, old name y49 new name y50
  In the expression
    (y : Type) -> y -> id Type y
warning: ./pi/test/IdTest.pi:6:19:
  generated a new renaming, old name _53 new name _54
  In the expression
    y -> id Type y
warning: ./pi/test/IdTest.pi:7:7:
  generated a new renaming, old name x75 new name x76
  In the expression
    \x y. y
warning: ./pi/test/IdTest.pi:7:7:
  generated a new renaming, old name y82 new name y83
  In the expression
    \x y. y
warning: After checking an entry there are unsolved constraints
With current active constraints being
{  127 : constraint 127: (_56 ~ Type) : Type
, 128 : constraint 128: (_56 ~ Type) : Type  }
And current solutions to metas being
  {  ?_1 : Type,
     ?_5 : Type,
     ?_9 : Type,
     ?_11 : Type,
     ?_13 : Type,
     ?_17 : Type,
     ?_18 : x16 -> x16,
     ?_19 : ?_31 [ ] -> ?_34 [ ],
     ?_24 : x16,
     ?_25 : x16,
     ?_26 : ?_40 [ ] -> ?_43 [ ],
     ?_29 : x16,
     ?_31 : Type,
     ?_34 : x16 -> x16,
     ?_40 : x16,
     ?_43 : x16,
     ?_47 : Type,
     ?_51 : Type,
     ?_55 : Type,
     ?_57 : _56 -> _56,
     ?_58 : ?_91 [ ] -> ?_94 [ ],
     ?_60 : Type,
     ?_62 : Type,
     ?_64 : Type,
     ?_65 : ?_117 [ ] -> ?_120 [ ],
     ?_67 : Type,
     ?_69 : Type,
     ?_71 : Type,
     ?_73 : Type,
     ?_77 : Type,
     ?_78 : x76 -> x76,
     ?_79 : ?_101 [ ] -> ?_104 [ ],
     ?_84 : x76,
     ?_85 : x76,
     ?_86 : ?_122 [ ] -> ?_125 [ ],
     ?_89 : x76,
     ?_91 : Type,
     ?_94 : _56 -> _56,
     ?_101 : Type,
     ?_104 : x76 -> x76,
     ?_122 : x76,
     ?_125 : x76  }
Solved constraints are
  {  2 : constraint 2: (Type ~ Type) : Type,
    6 : constraint 6: (Type ~ Type) : Type,
    10 : constraint 10: (Type ~ Type) : Type,
    12 : constraint 12: (Type ~ Type) : Type,
    14 : constraint 14: (Type ~ Type) : Type,
    20 : constraint 20: ((z : Type) -> z0 -> z0 ~ (x : ?_17 [ id0  →  id0 ]) -> ?_18 [ x16  →  x0, id0  →  id0 ]) : Type,
    27 : constraint 27: (?_18 [ id0  →  id0, x16  →  x16 ] ~ (y : ?_24 [ x16  →  x16, id0  →  id0 ]) -> ?_25 [ y23  →  y0, x16  →  x16, id0  →  id0 ]) : Type,
    30 : constraint 30: (?_24 [ id0  →  id0, x16  →  x16 ] ~ ?_25 [ id0  →  id0, x16  →  x16, y23  →  y23 ]) : Type,
    32 : constraint 32: (Type ~ ?_17 [ id0  →  id0 ]) : Type,
    35 : constraint 35: (z33 -> z33 ~ ?_18 [ x16  →  z33, id0  →  id0 ]) : Type,
    39 : constraint 39: (x16 -> x16 ~ (y : ?_24 [ x16  →  x16, id0  →  id0 ]) -> ?_25 [ y23  →  y0, x16  →  x16, id0  →  id0 ]) : Type,
    41 : constraint 41: (x16 ~ ?_24 [ x16  →  x16, id0  →  id0 ]) : Type,
    44 : constraint 44: (x16 ~ ?_25 [ y23  →  _42, x16  →  x16, id0  →  id0 ]) : Type,
    45 : constraint 45: (x16 ~ x16) : Type,
    48 : constraint 48: (Type ~ Type) : Type,
    52 : constraint 52: (Type ~ Type) : Type,
    59 : constraint 59: ((z : Type) -> z0 -> z0 ~ (_ : ?_55 [ _54  →  _54, y50  →  y50 ]) -> ?_57 [ _56  →  _0, _54  →  _54, y50  →  y50 ]) : Type,
    61 : constraint 61: (Type ~ ?_55 [ _54  →  _54, y50  →  y50 ]) : Type,
    66 : constraint 66: (?_57 [ _56  →  Type, _54  →  _54, y50  →  y50 ] ~ (_ : ?_62 [ _54  →  _54, y50  →  y50 ]) -> ?_64 [ _63  →  _0, _54  →  _54, y50  →  y50 ]) : Type,
    68 : constraint 68: (Type ~ ?_62 [ _54  →  _54, y50  →  y50 ]) : Type,
    70 : constraint 70: (?_64 [ _63  →  y50, _54  →  _54, y50  →  y50 ] ~ Type) : Type,
    72 : constraint 72: (Type ~ Type) : Type,
    74 : constraint 74: (Type ~ Type) : Type,
    80 : constraint 80: ((y : Type) -> y0 -> id0 Type y0 ~ (x : ?_77 [ id246  →  id246 ]) -> ?_78 [ x76  →  x0, id246  →  id246 ]) : Type,
    87 : constraint 87: (?_78 [ id246  →  id246, x76  →  x76 ] ~ (y : ?_84 [ x76  →  x76, id246  →  id246 ]) -> ?_85 [ y83  →  y0, x76  →  x76, id246  →  id246 ]) : Type,
    90 : constraint 90: (?_84 [ id246  →  id246, x76  →  x76 ] ~ ?_85 [ id246  →  id246, x76  →  x76, y83  →  y83 ]) : Type,
    92 : constraint 92: (Type ~ ?_55 [ _54  →  _54, y50  →  y50 ]) : Type,
    95 : constraint 95: (z93 -> z93 ~ ?_57 [ _56  →  z93, _54  →  _54, y50  →  y50 ]) : Type,
    102 : constraint 102: (Type ~ ?_77 [ id246  →  id246 ]) : Type,
    105 : constraint 105: (y103 -> id0 Type y103 ~ ?_78 [ x76  →  y103, id246  →  id246 ]) : Type,
    108 : constraint 108: (Type ~ Type) : Type,
    112 : constraint 112: (_56 -> _56 ~ (_ : ?_62 [ _54  →  _54, y50  →  y50 ]) -> ?_64 [ _63  →  _0, _54  →  _54, y50  →  y50 ]) : Type,
    116 : constraint 116: (x76 -> x76 ~ (y : ?_84 [ x76  →  x76, id246  →  id246 ]) -> ?_85 [ y83  →  y0, x76  →  x76, id246  →  id246 ]) : Type,
    118 : constraint 118: (_56 ~ ?_62 [ _54  →  _54, y50  →  y50 ]) : Type,
    121 : constraint 121: (_56 ~ ?_64 [ _63  →  _119, _54  →  _54, y50  →  y50 ]) : Type,
    123 : constraint 123: (x76 ~ ?_84 [ x76  →  x76, id246  →  id246 ]) : Type,
    126 : constraint 126: (x76 ~ ?_85 [ y83  →  _124, x76  →  x76, id246  →  id246 ]) : Type,
    129 : constraint 129: (x76 ~ x76) : Type  }
Blocked constraints are
  {   }
type checking...
Checking module "IdTest"
module IdTest where
id : (z : Type) -> z0 -> z0
id = \x y. y0
id2 : (y : Type) -> y0 -> id0 Type y0
id2 = \x y. y0
```

**Comment by liesnikov** [comment]
actually the problem was due to buggy reduction (fixed in cd4b983896d5eec5bdd720246122d9522e7c3ac0)
so e8bc9ef59a4b0ecf0b9b213195c8e7ce7db002ae was the fix
